libcontainer/factory_linux.go

```diff
@@ -372,6 +374,23 @@ func (l *LinuxFactory) StartInitialization() (err error) {
 		defer consoleSocket.Close()
 	}
 
+	// Get mount files (O_PATH)
+	mountFilesArr := strings.Split(envMountFileFds, ";")
+	mountFiles := make([]*os.File, len(mountFilesArr))
+	for i, fdStr := range mountFilesArr {
+		if fdStr == "" {
+			continue
+		}
+		mountFileFd, err := strconv.Atoi(fdStr)
+		if err != nil {
+			return fmt.Errorf("unable to parse _LIBCONTAINER_MOUNT_FILE_FDS(%q): %s",
+				envMountFileFds, err)
+		}
+		mountFile := os.NewFile(uintptr(mountFileFd), "mount-file")
+		defer mountFile.Close()
```

Comment:
These `defer`s will never be called because this function finished in `Execve`. There are two things we should do:

1. Explicitly set `O_CLOEXEC` on all of the file descriptors (`fcntl(F_SETFL, FD_CLOEXEC)` is what you want).
2. Close them as soon as we can.

Reply:
Thanks. I added a patch that does the following:
- the O_CLOEXEC is now set in `nsexec.c` by `dup3()`
- close them earlier in libcontainer/init_linux.go (for the initSetns case) and in libcontainer/standard_init_linux.go (for the initStandard case).

The "defer FOO.Close()" in this function follows the same code pattern of closing other fds in case `i.Init()` somehow fails to run `Execve`.